compiler/rustc_query_system/src/decl.rs

```diff
@@ -28,7 +28,7 @@ rustc_queries! {
 
     /// The indexed HIR. This can be conveniently accessed by `tcx.hir()`.
     /// Avoid calling this query directly.
-    query index_hir(_: ()) -> &'tcx crate::hir::IndexedHir<'tcx> {
+    query index_hir(_: ()) -> &'tcx rustc_middle::hir::IndexedHir<'tcx> {
```

Comment:
How does this work? `rustc_middle` has `extern crate rustc_query_system;` yet `rustc_query_system` imports a type from `rustc_middle`??? How is that not a cyclic dependency?

Reply:
All the rustc_queries macro does is to declare other macros. So from the point-of-view of the compiler, this is just some text, not a dependency to rustc_middle.

Reply:
Ah, that makes sense.

Reply:
BTW, this is why `can_reconstruct_query_key` is so painful.